### ISSUE TITLE
[dart2wasm] Async/microtask ordering is incorrect.

### ISSUE THREAD

**Issue by eyebrowsoffire** [dart-lang/sdk#50778]
When compiling with dart2wasm, the order in which microtasks resolve is not consistent with dart2wasm. This is exhibited in some of the tests in the `flutter/engine` repo under `flutter/lib/web_ui/test/channel_buffers_test.dart`.

**Comment by askeksa-google** [comment]
Is the issue that microtasks are expected to execute before other tasks in the event queue? If so, we might be able to work around it by executing all pending microtasks in these cases:
- Before executing a callback that is not a microtask (e.g. `Timer` callbacks)
- When we come out of an await (i.e. after calling `_futurePromise` in `_awaitHelper`)

This will naturally mess up the stack trace, so we'll need to fix those up somehow.

**Comment by eyebrowsoffire** [comment]
I found this issue cropping up in a few other engine tests. I produced a pretty minimal test case here that fails in dart2wasm but passes in dart2js:

```
import 'dart:async';

import 'package:async_helper/async_minitest.dart';

main() {
  test('interleaved', async () {
    final controller1 = StreamController<int>();
    final controller2 = StreamController<int>();
    final stream1 = controller1.stream.asBroadcastStream();
    final stream2 = controller2.stream.asBroadcastStream();
    
    controller1.add(1);
    controller2.add(2);

    expect(await stream1.first, equals(1));
    expect(await stream2.first, equals(2));
    
    controller1.add(3);
    controller2.add(4);
    
    expect(await stream1.first, equals(3));
    expect(await stream2.first, equals(4));
  });
}
```

**Comment by osa1** [comment]
@eyebrowsoffire is there a syntax error in your example? I think first line in main should be `() async {` instead of `async () {`, otherwise I'm getting this error:

```
test.dart:6:23: Error: A function expression can't have a name.
  test('interleaved', async () {
                      ^^^^^
```

With the change, I see this output in JIT:

```
unittest-suite-wait-for-done
unittest-suite-success
```

With dart2wasm the second line is not printed, but I don't see any exception failures.

#51599 is another and smaller task scheduling issue.

**Comment by lrhn** [comment]
Be careful with assuming something about "coming out of an `await`". Futures don't always use the microtask queue. A completing future can propagate its result to a number of transitive listeners without going back to the event loop.

Microtasks must run before any other event. The design, taken from the browser, is that every top-level event can trigger further microtasks, which then run in the context of that top-level event. When the microtask queue is empty, control goes back to the main event loop.
"In the context of" is significant on the web, where, e.g., a database transaction is a single event and the transaction gets committed when the event ends. You can still run more microtasks in the same context, before the transaction ends.

That said, we don't actually promise the precise timing of microtasks.

I'm actually surprised that `expect(await stream2.first, equals(4));` succeeds. I'd have expected that event to have been lost while the previous line is awaiting. Not code I'd ever promise a specific behavior of.

**Comment by osa1** [comment]
Micro-task scheduling should be fixed with 87a3de41b10d292d3069b2cca0cae13837669099 but `async*` compilation is still not quite right.

**Comment by mkustermann** [comment]
The minimal reproduction from https://github.com/dart-lang/sdk/issues/50778#issuecomment-1363387788 (when fixing the syntax error) succeeds in both dart2wasm and the VM.

The linked https://github.com/dart-lang/sdk/issues/51599 is also fixed.

Assumed stale.
